server/app/services/cloud/websocket_client.rb

```diff
@@ -91,91 +68,176 @@ def connecting?
       @connecting
     end
 
+    # @return [String]
+    def authorization_token
+      Base64.urlsafe_encode64(@client_id+':'+@client_secret)
+    end
+
     def connect
-      return if connecting?
-      if self.api_uri.to_s.empty?
-        error "Cloud Socket URI not configured"
-        return
-      end
-      @connected = false
       @connecting = true
+
+      info "Connecting to cloud at #{@api_uri}"
+
       headers = {
-        'Authorization' => "Basic #{Base64.urlsafe_encode64(self.client_id+':'+self.client_secret)}"
+        'Authorization' => "Basic #{authorization_token}"
       }
-      @ws = Faye::WebSocket::Client.new("#{self.api_uri}/platform", nil, { headers: headers })
+      @ws = Kontena::Websocket::Client.new(@api_uri,
+        headers: headers,
+        ssl_params: @ssl_params,
+        ssl_hostname: @ssl_hostname,
+        connect_timeout: CONNECT_TIMEOUT,
+        open_timeout: OPEN_TIMEOUT,
+        ping_interval: PING_INTERVAL,
+        ping_timeout: PING_TIMEOUT,
+        close_timeout: CLOSE_TIMEOUT,
+      )
+
+      async.connect_client @ws
 
-      @ws.on :open do |event|
-        on_open(event)
-      end
-      @ws.on :message do |event|
-        on_message(event)
-      end
-      @ws.on :close do |event|
-        on_close(event)
-      end
-      @ws.on :error do |event|
-        error "cloud connection closed with error: #{event.message}"
-      end
+    rescue => exc
+      error exc
 
+      # abort connect, allow re-connecting
+      @connecting = false
     end
 
-    ##
-    # @param [String, Array] msg
-    def send_message(msg)
-      EM.next_tick {
-        begin
-          @ws.send(msg) if @ws
-        rescue
-          error "failed to send message"
+    # Connect the websocket client, and read messages.
+    #
+    # Keeps running as a separate defer thread as long as the websocket client is connected.
+    #
+    # @param ws [Kontena::Websocket::Client]
+    def connect_client(ws)
+      actor = Actor.current
+
+      # run the blocking websocket client connect+read in a separate thread
+      defer {
+        # blocks until open, raises on errors
+        ws.connect
+
+        # These are called from the read_ws -> defer thread, proxy back to actor
+        actor.on_open
+
+        ws.read do |message|
+          actor.on_message(message)
         end
       }
+
+    rescue Kontena::Websocket::CloseError => exc
+      # handle known errors, will reconnect
+      on_close(exc.code, exc.reason)
+
+    rescue Kontena::Websocket::Error => exc
+      # handle known errors, will reconnect
+      on_close(1006, exc.message)
+
     rescue => exc
-      error "failed to send message: #{exc.message}"
+      # TODO: crash instead of reconnecting on unknown errors?
```

Comment:
IMO yes. Manager should be reported as they are linked and it can act on it.

Reply:
This is kinda TODO because there's still a lot of websocket-related socket/IO errors that the `kontena-websocket-client` doesn't trap: https://github.com/kontena/kontena-websocket-client/issues/1